File: 31-even-fib-sum/even_fib.go
// Golang program to find sum of all even fibonacci nums under 4 million
// Written by M.V.Harish Kumar on 29-05-2024
package main

import "fmt"

func even_fib_sum() int {
	sum := 0
	a, b := 0, 1
	for {
		c := a + b
		if c >= 4_000_000 {
			return sum
		}
		if (c % 2 == 0) {
			sum += c
		}
		a, b = b, c
	}
}

func main() {
	fmt.Println("Sum of all even fibonacci numbers under 4 million:", even_fib_sum())
}

File: 31-even-fib-sum/even_fib.exs
# Elixir program to find sum of all even fibonacci nums under 4 million
# Written by M.V.Harish Kumar on 06-02-2025

defmodule Fib do
  def even_fib_sum do
    Stream.from_index(1)
    |> Stream.map(&fib/1)
    |> Stream.take_while(&(&1 < 4_000_000))
    |> Enum.sum
  end

  def fibo(n), do: fib(n)

  defp fib(n), do: fib(n, 0, 1)
  defp fib(0, a, _), do: a
  defp fib(n, a, b), do: fib(n-1, a+b, a)
end

# IO.puts Fib.fibo(9)
IO.puts "Sum of all even fibonacci numbers under 4 million: #{Fib.even_fib_sum}"
